### PR TITLE
Httppoison to finch

### PR DESCRIPTION
**Pull request by njunghans**
Das ersetzen von httpoison zu finch sollte meinermeinung vollständig sein. Jedoch treten Fehler auf die meiner Meinung eher vom master abhängen, die httpoison vlcht geschlugt hat ?

Beim seeden : 
** (Mix) Could not start application argos_aggregation: exited in: ArgosAggregation.Application.start(:normal, [])
    ** (EXIT) an exception was raised:
        ** (CaseClauseError) no case clause matching: {:ok, %Finch.Response{body: "{\"error\":{\"root_cause\":[{\"type\":\"resource_already_exists_exception\",\"reason\":\"index [argos/XHMRZzLzSty2-Oc6CKTHRw] already exists\",\"index_uuid\":\"XHMRZzLzSty2-Oc6CKTHRw\",\"index\":\"argos\"}],\"type\":\"resource_already_exists_exception\",\"reason\":\"index [argos/XHMRZzLzSty2-Oc6CKTHRw] already exists\",\"index_uuid\":\"XHMRZzLzSty2-Oc6CKTHRw\",\"index\":\"argos\"},\"status\":400}", headers: [{"content-type", "application/json; charset=UTF-8"}, {"content-length", "361"}], status: 400}}
            (argos_aggregation 0.1.0) lib/argos_aggregation/application.ex:55: ArgosAggregation.Application.initialize_index/0
            (argos_aggregation 0.1.0) lib/argos_aggregation/application.ex:96: ArgosAggregation.Application.start/2
            (kernel 6.4.1) application_master.erl:277: :application_master.start_it_old/4

Beim testen : 
** (FunctionClauseError) no function clause matching in Finch.Request.parse_url/1

     The following arguments were given to Finch.Request.parse_url/1:
     
         # 1
         'http://thesauri.dainst.org/search.rdf?change_note_date_from=2021-07-04&page=1&q='
     
     Attempted function clauses (showing 2 out of 2):
     
         def parse_url(url)         when is_binary(url)
         def parse_url(%URI{} = parsed_uri)
     
     code: |> Harvester.run_harvest
     stacktrace:
       (finch 0.7.0) Finch.Request.parse_url/1
       (finch 0.7.0) lib/finch/request.ex:60: Finch.Request.build/4
       (argos_aggregation 0.1.0) lib/argos_aggregation/data_sources/thesauri.ex:52: ArgosAggregation.Thesauri.DataSourceClient.Http.read_from_url/1
       (argos_aggregation 0.1.0) lib/argos_aggregation/data_sources/thesauri.ex:120: ArgosAggregation.Thesauri.DataProvider.load_next_page/1
       (elixir 1.13.0-dev) lib/stream.ex:1531: Stream.do_resource/5
       (elixir 1.13.0-dev) lib/stream.ex:1719: Enumerable.Stream.do_each/4
       (elixir 1.13.0-dev) lib/enum.ex:3920: Enum.each/2
       test/data_sources/thesauri_test.exs:160: (test)

### CODE REVIEW

**Comment by njunghans** [comment]
Hab die ungültige url gefunden
defp stream_pages(date) do
      Stream.resource(
        fn ->
          case DataSourceClient.Http.request_by_date(date) do
            {:ok, xml} -> ConceptParser.Search.load_first_page_url(xml)
            error -> {[error], nil}
          end
        end,
        fn page_url ->
          case page_url do
            nil -> {:halt, page_url}
            page_url -> load_next_page(page_url|>to_string())

**Comment by preiaen** [comment]
> Das ersetzen von httpoison zu finch sollte meinermeinung vollständig sein. Jedoch treten Fehler auf die meiner Meinung eher vom master abhängen, die httpoison vlcht geschlugt hat ?
> 
> Beim seeden :
> ** (Mix) Could not start application argos_aggregation: exited in: ArgosAggregation.Application.start(:normal, [])
> ** (EXIT) an exception was raised:
> ** (CaseClauseError) no case clause matching: {:ok, %Finch.Response{body: "{"error":{"root_cause":[{"type":"resource_already_exists_exception","reason":"index [argos/XHMRZzLzSty2-Oc6CKTHRw] already exists","index_uuid":"XHMRZzLzSty2-Oc6CKTHRw","index":"argos"}],"type":"resource_already_exists_exception","reason":"index [argos/XHMRZzLzSty2-Oc6CKTHRw] already exists","index_uuid":"XHMRZzLzSty2-Oc6CKTHRw","index":"argos"},"status":400}", headers: [{"content-type", "application/json; charset=UTF-8"}, {"content-length", "361"}], status: 400}}
> (argos_aggregation 0.1.0) lib/argos_aggregation/application.ex:55: ArgosAggregation.Application.initialize_index/0
> (argos_aggregation 0.1.0) lib/argos_aggregation/application.ex:96: ArgosAggregation.Application.start/2
> (kernel 6.4.1) application_master.erl:277: :application_master.start_it_old/4
> 
> Beim testen :
> ** (FunctionClauseError) no function clause matching in Finch.Request.parse_url/1
> 
> ```
>  The following arguments were given to Finch.Request.parse_url/1:
>  
>      # 1
>      'http://thesauri.dainst.org/search.rdf?change_note_date_from=2021-07-04&page=1&q='
>  
>  Attempted function clauses (showing 2 out of 2):
>  
>      def parse_url(url)         when is_binary(url)
>      def parse_url(%URI{} = parsed_uri)
>  
>  code: |> Harvester.run_harvest
>  stacktrace:
>    (finch 0.7.0) Finch.Request.parse_url/1
>    (finch 0.7.0) lib/finch/request.ex:60: Finch.Request.build/4
>    (argos_aggregation 0.1.0) lib/argos_aggregation/data_sources/thesauri.ex:52: ArgosAggregation.Thesauri.DataSourceClient.Http.read_from_url/1
>    (argos_aggregation 0.1.0) lib/argos_aggregation/data_sources/thesauri.ex:120: ArgosAggregation.Thesauri.DataProvider.load_next_page/1
>    (elixir 1.13.0-dev) lib/stream.ex:1531: Stream.do_resource/5
>    (elixir 1.13.0-dev) lib/stream.ex:1719: Enumerable.Stream.do_each/4
>    (elixir 1.13.0-dev) lib/enum.ex:3920: Enum.each/2
>    test/data_sources/thesauri_test.exs:160: (test)
> ```

Dieses Problem scheint sich erledigt zu haben richtig? Die Tests laufen bei mir alle durch.

**Review by preiaen — COMMENTED** [review]
Also insgesamt sieht das so aus, als wenn man das so lassen kann. 
Allerdings denke ich nach wie vor, dass es sinnvoll gewesen wäre, eine Wrapper Modul einzusetzen. Diese enge Bindung an externe Apis ist einfach nicht gut und dieser Fehler, den du hattest, hätte man da auch vermeiden können. Warum soll jedes Modul selber auf HTTPoioson.response oder Finch.response testen, wenn das in einem Modul zentral gemacht werden könnte?

File: apps/argos_aggregation/lib/argos_aggregation/application.ex [hunk 1]
```diff
@@ -31,45 +32,54 @@ defmodule ArgosAggregation.Application do
...
+    Finch.build(:put, "#{@elasticsearch_url}/_mapping", [{"Content-Type", "application/json"}],mapping)
+    |> Finch.request(ArgosFinch)
   end
 
 
   defp initialize_index() do
-    case HTTPoison.get(@elasticsearch_url) do
+    res =Finch.build(:put, @elasticsearch_url)
+    |> Finch.request(ArgosFinch)
+    case res do
```
Comment (preiaen) [inline apps/argos_aggregation/lib/argos_aggregation/application.ex:55]: Das seeding Problem entsteht in diesem Case:
Der Endpoint Antwortet mit 
`{:ok, %Finch.Response{body: "{\"error\":{\"root_cause\":[{\"type\":\"resource_already_exists_exception\"...`
Dafür gibt es aber keinen Case Fall und damit steigt Argos mit dem Fehler 
`(CaseClauseError) no case clause matching:`
aus.

Das case braucht glaube ich insgesamt noch etwas liebe, es ist nur sehr schwer da durchzusteigen.

**Review by preiaen — CHANGES_REQUESTED** [review]
Das case sollte vielleicht nochmal angepasst werden. s. o.

File: apps/argos_aggregation/lib/argos_aggregation/application.ex (hunk 1 above)
Comment (njunghans) [reply apps/argos_aggregation/lib/argos_aggregation/application.ex:55]: aber warum hat es mit httpoison davor funktionniert ?
da gab es auch schon kein case für "already exists"

File: apps/argos_aggregation/lib/argos_aggregation/application.ex [hunk 2]
```diff
@@ -31,45 +32,54 @@ defmodule ArgosAggregation.Application do
...
-    "#{@elasticsearch_url}/_mapping"
-    |> HTTPoison.put(mapping, [{"Content-Type", "application/json"}])
+    Finch.build(:put, "#{@elasticsearch_url}/_mapping", [{"Content-Type", "application/json"}],mapping)
+    |> Finch.request(ArgosFinch)
   end
 
 
   defp initialize_index() do
-    case HTTPoison.get(@elasticsearch_url) do
+    res =Finch.build(:put, @elasticsearch_url)
```
Comment (preiaen) [inline apps/argos_aggregation/lib/argos_aggregation/application.ex:53]: Du änderst hier die Anfrage von GET auf PUT, vmtl reagiert Elastic deshalb anders.

**Review by dersmon — CHANGES_REQUESTED** [review]
Das mit der Vereinheitlichung können wir auch in ein neues Ticket auslagern? Ich bin mir nicht 100% sicher wieviel dadurch gewonnen wird.

File: apps/argos_aggregation/lib/argos_aggregation/data_sources/thesauri.ex [hunk 3]
```diff
@@ -107,7 +107,7 @@ defmodule ArgosAggregation.Thesauri do
         fn page_url ->
           case page_url do
             nil -> {:halt, page_url}
-            page_url -> load_next_page(page_url)
+            page_url -> load_next_page(page_url|>to_string())
```
[the remark this answers is not in the document]
Comment (njunghans) [reply apps/argos_aggregation/lib/argos_aggregation/data_sources/thesauri.ex:110]: ** (FunctionClauseError) no function clause matching in Finch.Request.parse_url/1

     The following arguments were given to Finch.Request.parse_url/1:
     
         # 1
         'http://thesauri.dainst.org/search.rdf?change_note_date_from=2021-07-30&page=1&q='
     
     Attempted function clauses (showing 2 out of 2):
     
         def parse_url(url)         when is_binary(url)
         def parse_url(%URI{} = parsed_uri)
         
        Kommt falls dies kein string ist.

Habe folgendes benutzt:
defmodule Util do
    types = ~w[function nil integer binary bitstring list map float atom tuple pid port reference]
    for type <- types do
      def typeof(x) when unquote(:"is_#{type}")(x), do: unquote(type)
    end
  end
  
  um rauszufinden das es eine liste ist.

File: apps/argos_aggregation/test/data_sources/chronontology_test.exs [hunk 4]
```diff
@@ -30,7 +30,7 @@ defmodule ArgosAggregation.ChronontologyTest do
   test "get by id with invalid id yields 404" do
     id = "i-am-non-existant"
 
-    expected_result = {:error, "Received unhandled status code 404 for https://chronontology.dainst.org/data/period/#{id}."}
+    expected_result = {:error, "Received unhandled status code 404."}
```
[the remark this answers is not in the document]
Comment (njunghans) [reply apps/argos_aggregation/test/data_sources/chronontology_test.exs:33]: kein hack. Ich habe die responses der provider ja geändert. Deshalb mussten die Tests auch verändert werden.
Grund ist : bei Finch.Response ist der request nicht dabei (https://hexdocs.pm/finch/Finch.Response.html#content) wie bei Poison. Deshalbt habe ich das feature rausgenommen.

File: apps/argos_aggregation/lib/argos_aggregation/application.ex [hunk 5]
```diff
@@ -81,6 +90,7 @@ defmodule ArgosAggregation.Application do
   end
 
   def start(_type, _args) do
+    Finch.start_link(name: ArgosFinch)
```
[the remark this answers is not in the document]
Comment (njunghans) [reply apps/argos_aggregation/lib/argos_aggregation/application.ex:93]: mix seeds.projects braucht Finch auch, deshalb so

[the remark this answers is not in the document]
Comment (njunghans) [reply apps/argos_aggregation/lib/argos_aggregation/application.ex:93]: eine mögliche anpassung ?
`children =
      if running_script?(System.argv) do
        [] # We do not want to (re)start the harvesters when running exs scripts.
      else
        @active_harvesters
      end

    if Application.get_env(:argos_aggregation, :await_index, true) do
      Finch.start_link(name: ArgosFinch)
      initialize_index()
    end

    children = if !Application.get_env(:argos_aggregation, :await_index, true) do [{Finch, name: ArgosFinch}] ++ children else children end

    # See https://hexdocs.pm/elixir/Supervisor.html
    # for other strategies and supported options
    opts = [strategy: :one_for_one, name: ArgosAggregation.Supervisor]

    Supervisor.start_link(children, opts)`

File: apps/argos_aggregation/lib/argos_aggregation/data_sources/thesauri.ex (hunk 3 above)
[the remark this answers is not in the document]
Comment (njunghans) [reply apps/argos_aggregation/lib/argos_aggregation/data_sources/thesauri.ex:110]: aber mit to_string() wird es zu einer gültigen url

File: apps/argos_aggregation/test/data_sources/chronontology_test.exs (hunk 4 above)
[the remark this answers is not in the document]
Comment (dersmon) [reply apps/argos_aggregation/test/data_sources/chronontology_test.exs:33]: Ah, macht Sinn.

File: apps/argos_aggregation/lib/argos_aggregation/data_sources/thesauri.ex (hunk 3 above)
[the remark this answers is not in the document]
Comment (dersmon) [reply apps/argos_aggregation/lib/argos_aggregation/data_sources/thesauri.ex]: Verstehe, an den einfachen Anführungszeichen kann man sehen, dass SweetXML hier anscheinend eine charlist anstatt eines "normalen" binary strings (das wären doppelte Anführungszeichen) zurückgibt. Im Gegensatz zu HTTPoison scheint Finch char lists nicht nutzen zu können. Deshalb warst du hier zum `to_string` gezwungen auf einmal.

Ich habe mal einen Kommentar eingefügt.

[the remark this answers is not in the document]
Comment (dersmon) [reply apps/argos_aggregation/lib/argos_aggregation/data_sources/thesauri.ex]: Siehe: https://elixir-lang.org/getting-started/binaries-strings-and-char-lists.html

File: apps/argos_aggregation/lib/argos_aggregation/application.ex [hunk 6]
```diff
@@ -95,7 +100,14 @@ defmodule ArgosAggregation.Application do
     # See https://hexdocs.pm/elixir/Supervisor.html
     # for other strategies and supported options
     opts = [strategy: :one_for_one, name: ArgosAggregation.Supervisor]
+    children = [{Finch, name: ArgosAggregationFinch}] ++ children
```
Comment (preiaen) [inline apps/argos_aggregation/lib/argos_aggregation/application.ex:103]: Die Bezeichnung ArgosAggregationFinch ist hier etwas unglücklich. Es besteht die Gefahr es mit einem eigenständigen Modul zu verwechseln, da weder Schreibweise noch der Text erkennen lassen, dass es sich um einen child process handelt.

**Review by preiaen — COMMENTED** [review]
Abgesehen von dem bereits diskutierten Namen des ArgosAggregationFinchProcess sieht alles gut aus.

**Review by dersmon — APPROVED** [review]

**Pull request merged**